pkg/controller/direct/sql/merge.go

```diff
@@ -0,0 +1,906 @@
+			ReuseInterval:             direct.ValueOf(desired.Spec.Settings.PasswordValidationPolicy.ReuseInterval),
+		}
+	} else if actual.Settings.PasswordValidationPolicy != nil {
+		// Remove password validation policy
+		updateRequired = true
+	}
+
+	if desired.Spec.Settings.PricingPlan != nil {
+		if actual.Settings.PricingPlan == "" {
+			// Add pricing plan
+			updateRequired = true
+		} else if direct.ValueOf(desired.Spec.Settings.PricingPlan) != actual.Settings.PricingPlan {
+			// Change pricing plan
+			updateRequired = true
+		}
+		merged.Settings.PricingPlan = direct.ValueOf(desired.Spec.Settings.PricingPlan)
+	} else if actual.Settings.PricingPlan != "" {
+		// Keep pricing plan
+		merged.Settings.PricingPlan = actual.Settings.PricingPlan
+	}
+
+	if desired.Spec.Settings.ReplicationType != nil {
+		if actual.Settings.ReplicationType == "" {
+			// Add replication type
+			updateRequired = true
+		} else if direct.ValueOf(desired.Spec.Settings.ReplicationType) != actual.Settings.ReplicationType {
+			// Change replication type
+			updateRequired = true
+		}
+		merged.Settings.ReplicationType = direct.ValueOf(desired.Spec.Settings.ReplicationType)
+	} else if actual.Settings.ReplicationType != "" {
+		// Keep replication type
+		merged.Settings.ReplicationType = actual.Settings.ReplicationType
+	}
+
+	if desired.Spec.Settings.SqlServerAuditConfig != nil {
+		if actual.Settings.SqlServerAuditConfig == nil {
+			// Add sql server audit config
+			updateRequired = true
+		} else if (desired.Spec.Settings.SqlServerAuditConfig.BucketRef.Name != actual.Settings.SqlServerAuditConfig.Bucket) ||
+			(direct.ValueOf(desired.Spec.Settings.SqlServerAuditConfig.RetentionInterval) != actual.Settings.SqlServerAuditConfig.RetentionInterval) ||
+			(direct.ValueOf(desired.Spec.Settings.SqlServerAuditConfig.UploadInterval) != actual.Settings.SqlServerAuditConfig.UploadInterval) {
+			// Change sql server audit config
+			updateRequired = true
+		}
+		merged.Settings.SqlServerAuditConfig = &api.SqlServerAuditConfig{
+			Bucket:            desired.Spec.Settings.SqlServerAuditConfig.BucketRef.Name,
+			RetentionInterval: direct.ValueOf(desired.Spec.Settings.SqlServerAuditConfig.RetentionInterval),
+			UploadInterval:    direct.ValueOf(desired.Spec.Settings.SqlServerAuditConfig.UploadInterval),
+			Kind:              "sql#sqlServerAuditConfig",
+		}
+	} else if actual.Settings.SqlServerAuditConfig != nil {
+		// Remove sql server audit config
+		updateRequired = true
+	}
+
+	if desired.Spec.Settings.Tier != "" {
+		if actual.Settings.Tier == "" {
+			// Add tier
+			updateRequired = true
+		} else if desired.Spec.Settings.Tier != actual.Settings.Tier {
+			// Change tier
+			updateRequired = true
+		}
+		merged.Settings.Tier = desired.Spec.Settings.Tier
+	} else {
+		// Keep tier
+		merged.Settings.Tier = actual.Settings.Tier
+	}
+
+	if desired.Spec.Settings.TimeZone != nil {
```

Comment:
IIUC, `settings.timeZone` is actually an immutable field, so I assume we actually want to return an error here?

I'm not sure if I've missed any context about how we plan to block the update of the immutable fields in SciFi resources. Do you plan to add the immutable field validator in the webhook?

Reply:
I haven't gotten around to testing all the update scenarios yet. Will probably handle this in a future PR. It would probably be best ux to use a webhook to validate these immutable fields do not change, and maybe double-check in the reconciler also.